src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ActivatorUtilities.cs

```diff
@@ -105,18 +130,94 @@ public static class ActivatorUtilities
                 }
             }
 
-            Type?[] argumentTypes = new Type[parameters.Length];
-            for (int i = 0; i < argumentTypes.Length; i++)
+            Type?[] argumentTypes;
+            if (parameters.Length == 0)
+            {
+                argumentTypes = Type.EmptyTypes;
+            }
+            else
             {
-                argumentTypes[i] = parameters[i]?.GetType();
+                argumentTypes = new Type[parameters.Length];
+                for (int i = 0; i < argumentTypes.Length; i++)
+                {
+                    argumentTypes[i] = parameters[i]?.GetType();
+                }
             }
 
             FindApplicableConstructor(instanceType, argumentTypes, out ConstructorInfo constructorInfo, out int?[] parameterMap);
-            var constructorMatcher = new ConstructorMatcher(constructorInfo);
+
+            // Find the ConstructorInfoEx from the given constructorInfo.
+            constructor = null;
+            foreach (ConstructorInfoEx ctor in constructors)
+            {
+                if (ReferenceEquals(ctor.Info, constructorInfo))
+                {
+                    constructor = ctor;
+                    break;
+                }
+            }
+
+            Debug.Assert(constructor != null);
+
+            var constructorMatcher = new ConstructorMatcher(constructor);
             constructorMatcher.MapParameters(parameterMap, parameters);
             return constructorMatcher.CreateInstance(provider);
         }
 
+#if NETCOREAPP
+        private static ConstructorInfoEx[] GetOrAddConstructors(
+            [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors)] Type type)
+        {
+            // Not found. Do the slower work of checking for the value in the correct cache.
+            // Null and non-collectible load contexts use the default cache.
+            if (!type.Assembly.IsCollectible)
+            {
+                return s_constructorInfos.GetOrAdd(type, CreateConstructorInfoExs(type));
+            }
+
+            // Collectible load contexts should use the ConditionalWeakTable so they can be unloaded.
+            if (s_collectibleConstructorInfos.Value.TryGetValue(type, out ConstructorInfoEx[]? value))
+            {
+                return value;
+            }
+
+            value = CreateConstructorInfoExs(type);
+
+#if NET7_0_OR_GREATER
+            // ConditionalWeakTable doesn't support GetOrAdd(), so use TryAdd().
+            if (s_collectibleConstructorInfos.Value.TryAdd(type, value))
+            {
+                return value;
+            }
+#else
+            try
+            {
+                s_collectibleConstructorInfos.Value.Add(type, value);
+                return value;
+            }
+            catch (ArgumentException) { }
```

Comment:
`ConditionalWeakTable.Add(...)` would throw ArgumentException if the key/value not added (because its already added with other thread)?  NIT: Maybe add a comment for why ignoring the `ArgumentException`

Reply:
Yes that is the case. I simplified this to just use AddOrUpdate() -- previously the code used the same instance for race conditions when adding to the cache, but I don't think that was worth it, and could possibly lead to an issue with hot reload which could clear the cache after the exception but before the TryGet().